File: BAEKJOON/data_structure/javascript/1406.js
const fs = require("fs");
let input =
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString().trim()
    : fs
        .readFileSync(__dirname + "/input.txt")
        .toString()
        .trim();

input = input.split("\n");
const initString = input.shift();
const n = +input.shift();
input = input.map((command) => command.split(" "));

function solution(input, init) {
  const leftStack = [...init.split("")];
  const rightStack = [];

  for (let command of input) {
    if (command[0] === "L" && leftStack.length !== 0) {
      rightStack.push(leftStack.pop());
    }

    if (command[0] === "D" && rightStack.length !== 0) {
      leftStack.push(rightStack.pop());
    }

    if (command[0] === "B") {
      leftStack.pop();
    }

    if (command[0] === "P") {
      leftStack.push(command[1]);
    }
  }

  return leftStack.concat(rightStack.reverse()).join("");
}

console.log(solution(input, initString));

File: BAEKJOON/data_structure/python/1406.py
""" 
22.07.07
BOJ : 1406 / 자료구조
"""
import sys

left = list(sys.stdin.readline().strip())
right = []
n = int(input())

for _ in range(n):
    cmds = sys.stdin.readline().strip().split(' ')
    c = cmds[0]

    if c == 'P':
        left.append(cmds[1])
    elif c == 'L' and left != []:
        right.append(left.pop())
    elif c == 'D' and right != []:
        left.append(right.pop())
    elif c == 'B' and left != []:
        left.pop()

print(''.join(left + right[::-1]))

"""
스택을 사용하지 않고 insert, pop(index)를 사용 시 시간복잡도가 크게 증가
"""
